crates/parser/src/ast.rs

```diff
@@ -439,3 +769,93 @@ impl fmt::Display for CompOperator {
         }
     }
 }
+
+fn node_comma_joined(nodes: &[Node<impl fmt::Display>]) -> String {
+    comma_joined(&nodes.iter().map(|node| &node.kind).collect::<Vec<_>>())
+}
+
+fn comma_joined(items: &[impl fmt::Display]) -> String {
+    items
+        .iter()
+        .map(|item| format!("{}", item))
+        .collect::<Vec<_>>()
+        .join(", ")
+}
+
+fn node_line_joined(nodes: &[Node<impl fmt::Display>]) -> String {
+    line_joined(&nodes.iter().map(|node| &node.kind).collect::<Vec<_>>())
+}
+
+fn line_joined(items: &[impl fmt::Display]) -> String {
+    items
+        .iter()
+        .map(|item| format!("{}", item))
+        .collect::<Vec<_>>()
+        .join("\n")
+}
+
+fn maybe_fmt_inner_with_parens(outer: &Expr, inner: &Expr) -> String {
+    if inner_expr_needs_parens(outer, inner) {
+        format!("({})", inner)
+    } else {
+        format!("{}", inner)
+    }
+}
+
+/// An inner expression needs parentheses if the power of the outer expression is greater than
+/// that of the inner.
+///
+/// e.g.
+/// In the expression `2 * (3 + 4)`, the power of an `*` operation is greater than the power of
+/// an `+` operation. If we do not account for this by wrapping the `+` operation in parentheses,
+/// then the expression changes meaning.
+fn inner_expr_needs_parens(outer: &Expr, inner: &Expr) -> bool {
+    expr_power(outer) > expr_power(inner)
+}
```

Comment:
The paren insertion logic isn't right. Eg the parens of `2 * (3 / 4)` are dropped. For left-associative ops, using `>` for the lhs of a binop, and `>=` for the rhs would suffice, but `**` is right-associative, so `(2 ** 3) ** 4` needs to keep its parens but `2 ** (3 ** 4)` doesn't.

Maybe there's a more elegant way, but I think of it as left and right binding power, ie the amount of attraction power an op has on its left and right operands, as in `infix_binding_power`. A single `expr_power` fn isn't sufficient, as the power an outer expr exerts on its lhs and rhs might be different, and the power an inner expr   responds with may differ depending on whether it's on the left or right.

Reply:
Ok, updated it to use left and right binding power. I think it's correct now.